### PR TITLE
Generate a list of appOptions objects when there are multiple levels to choose from

### PR DESCRIPTION
**Pull request by balderdash**
Currently restricting level variants to have the same game type, so that I don't have to deal with differing sets of javascript dependencies.

Mostly, this change just replaces uses of `@level` with a level parameter passed into the function instead, so that I can iterate over `@levels` and generate appOptions for each.

https://docs.google.com/document/d/1lIL3PCVnubDc8srxc5p8UKqu6jNl60qXH0y4rloV-kw/edit#heading=h.p6ulezkf5j1d

### CODE REVIEW

**Comment by Hamms** [comment]
Is this ready for review? Those test failures look legitimate

**Comment by balderdash** [comment]
Oh oops. Yeah that's a problem, un-assigning you for now.

File: dashboard/app/controllers/script_levels_controller.rb [hunk 1]
```diff
@@ -199,32 +199,34 @@ def load_section
     end
   end
 
-  def select_level
+  def select_levels
     # If there's only one level in this scriptlevel, use that
-    return @script_level.levels[0] if @script_level.levels.length == 1
+    return @script_level.levels if @script_level.levels.length == 1
 
-    # For teachers, load the student's most recent attempt
```
Comment (Hamms) [inline dashboard/app/controllers/script_levels_controller.rb:206]: why was this removed?

File: dashboard/app/controllers/script_levels_controller.rb [hunk 2]
```diff
@@ -199,32 +199,34 @@ def load_section
...
-    if @user && current_user != @user
-      last_attempt = @user.last_attempt_for_any(@script_level.levels)
-      return last_attempt.level if last_attempt
-    end
-
-    # If they've tried at least one variant before, use the most recently attempted
-    # (unless overridden by a force_reload query string param)
+    # If their most recent attempt was on a now-inactive level, show that.
```
Comment (Hamms) [inline dashboard/app/controllers/script_levels_controller.rb:209]: why only consider most recent attempt for inactive levels?

File: dashboard/app/helpers/levels_helper.rb [hunk 3]
```diff
@@ -50,14 +50,14 @@ def readonly_view_options
     view_options callouts: []
   end
 
-  def set_channel
+  def set_channel(level)
```
Comment (Hamms) [inline dashboard/app/helpers/levels_helper.rb:53]: any reason to not just do `set_channel(level = @level)`?

File: dashboard/app/helpers/levels_helper.rb [hunk 4]
```diff
@@ -521,14 +500,6 @@ def applab_user_id
...
   end
 
-  def enable_scrolling?
-    @level.is_a?(Blockly)
-  end
-
-  def enable_examples?
-    @level.is_a?(Blockly)
-  end
-
```
Comment (Hamms) [inline dashboard/app/helpers/levels_helper.rb:531]: nice cleanup!

**Comment by Hamms** [comment]
This seems like it should be reviewed by someone a bit more intimately familiar with the intricacies of appOptions generation than I am. Josh or Dave, perhaps?

File: dashboard/app/controllers/script_levels_controller.rb (hunk 1 above)
Comment (balderdash) [inline dashboard/app/controllers/script_levels_controller.rb:206]: The level selection logic for teachers is now the same as it is for students, but I forgot about the difference between `@user` and current_user. Added a line back to handle that.

File: dashboard/app/controllers/script_levels_controller.rb (hunk 2 above)
Comment (balderdash) [inline dashboard/app/controllers/script_levels_controller.rb:209]: The selector won't show an inactive level, so if their most recent activity is on an inactive level, that has to be the only level returned to the client.

If there are multiple active levels (and there are no inactive levels/their most recent activity was not on the inactive level), we'll be showing the selector anyway, so no need to pick just the most recent.

File: dashboard/app/helpers/levels_helper.rb (hunk 3 above)
Comment (balderdash) [inline dashboard/app/helpers/levels_helper.rb:53]: Basically to fail fast. I don't want it to silently use the wrong level if I missed a set_channel call.

**Comment by balderdash** [comment]
@davidsbailey Are you more intimately familiar with the way `appOptions` is generated?

**Comment by davidsbailey** [comment]
Looking... heads up that github says the branch has merge conflicts

File: code-studio/src/js/initApp/initApp.js [hunk 5]
```diff
@@ -19,6 +19,13 @@ window.apps = {
   // Loads the dependencies for the current app based on values in `appOptions`.
   // This function takes a callback which is called once dependencies are ready.
   load: loadApp,
+  chooseApp: function (selectorProperties, appOptionsList) {
+    if (appOptionsList.size === 1) {
```
Comment (joshlory) [inline code-studio/src/js/initApp/initApp.js:23]: I think you want `.length` here.

File: dashboard/app/helpers/levels_helper.rb (hunk 3 above)
Comment (davidsbailey) [inline dashboard/app/helpers/levels_helper.rb:53]: @balderdash  I think I see what you are saying here, but I am concerned that this change makes the LevelsHelper class harder to understand. I think we need to choose between:
1. a helper class which operates on `@level`, except when explicitly overridden (i.e. Elijah's comment)
2. a helper class which is initialized with a level, which then operates on that level and whose methods don't take a level param (not sure how well this works in the rails / controller / helper world)

My reasoning is that it's confusing that the `level` param is required param in almost every method, but that `@level` is still referenced directly in some methods. And if you're going to require the `level` param in every method, then it would be better to supply it when you initialize the class.

File: code-studio/src/js/initApp/initApp.js [hunk 6]
```diff
@@ -19,6 +19,13 @@ window.apps = {
   // Loads the dependencies for the current app based on values in `appOptions`.
   // This function takes a callback which is called once dependencies are ready.
   load: loadApp,
+  chooseApp: function (selectorProperties, appOptionsList) {
```
Comment (davidsbailey) [inline code-studio/src/js/initApp/initApp.js:22]: Please explain in a function-level comment what `selectorProperties` is and how you will use it to select the `appOptions`

**Comment by joshlory** [comment]
The cleanup looks good.  I wonder if it would make things easier to have an interstitial page for choosing a variation, before deciding which version of `appOptions` to send to the client.

On pages that have lots of variations we could end up sending down lots of `appOptions` versions.  We'll still need the `appOptions` config even if the level isn't marked `active` right?  But the JSON isn't that big so this isn't a huge concern.

**Comment by davidsbailey** [comment]
On a particular applab level, it looks like there is about 20KB of level-specific data and another 20KB of non-level-specific data: https://studio.code.org/s/allthethings/stage/18/puzzle/10

```
JSON.stringify(appOptions).length
40922
JSON.stringify(appOptions.level).length
17067
```

This is something you'll probably want to address before you launch this feature, by moving any level-specific things you need into appOptions.level and then only duplicating that part of appOptions. If you know you're going to need to do this, it's probably easier to do sooner than later, but it doesn't necessarily have to be in this PR.

File: dashboard/app/helpers/levels_helper.rb [hunk 7]
```diff
@@ -170,30 +171,32 @@ def app_options
...
+    elsif level.is_a? Widget
+      @app_options[level.id] = widget_options level
+    elsif level.unplugged?
+      @app_options[level.id] = unplugged_options
     else
       # currently, all levels are Blockly or DSLDefined except for Unplugged
-      @app_options = view_options.camelize_keys
+      @app_options[level.id] = view_options.camelize_keys
     end
+    reset_view_options
```
Comment (davidsbailey) [inline dashboard/app/helpers/levels_helper.rb:186]: A few comments related to the freeze / reset:
- This seems sketchy to me because you are resetting and regenerating the view options, but whichever level you load last is going to get cached as `@view_options` and referenced by things like `@view_options.readonly_workspace` in `_single_text_match.html.haml`. If your levels differ on that property, you would get a wrong result.
- https://www.pivotaltracker.com/n/projects/914728/stories/98153794 says that we simply shouldn't be freezing view options in the first place
- perhaps if we are setting appOptions.level as an array `appOptions.levels` this problem will become moot

Comment (davidsbailey) [inline dashboard/app/helpers/levels_helper.rb:186]: one incremental change to make this saner would be to change `view_options()` (no arguments) to `freeze_view_options` so that it's obvious when it's getting frozen vs. being called in a non-freezing manner

**Comment by davidsbailey** [comment]
Hey @balderdash , nice progress in an area that obviously holds a lot of technical debt. I think the main concern for me is any correctness issues related to referencing `@view_options` which will always be set to whichever level was generated last. Let's discuss in person tomorrow which of the other things I commented on it makes sense to do now, do later or not do.

**Comment by balderdash** [comment]
Re: Size of appOptions

During javascript initialization, a bunch of stuff gets added to appOptions. It starts out a lot smaller, with 1.5KB of level options and another ~4.5KB of app options:

JSON.stringify(appOptions.level).length
1522
JSON.stringify(appOptions).length
5860

For reference, loading https://studio.code.org/s/allthethings/stage/18/puzzle/10 took about 1.5MB total.

I still think it's worth duplicating just the level options instead of all the appOptions, but the download size doesn't seem like a major issue.
